### ISSUE TITLE
Unexpected results in LINQs ExceptBy

### ISSUE THREAD

**Issue by AbstractionsAs** [dotnet/runtime#93512]
### Description

Given the following list:

```csharp
var list = new[]
{
	new { Name = "Should be returned 1", State = "Enabled" },
	new { Name = "Should be returned 2", State = "Bob" },
	new { Name = "Should be returned 3", State = "" },
	new { Name = "Should be returned 4", State = default(string) },
	new { Name = "Should be filtered 1", State = "disAbled" },
	new { Name = "Should be returned 5", State = "Enabled" },
};
```

and the following filter:
```csharp
var filter = new[] { "Disabled", "Deleted", "Warned" };
```

I would expect the following statements to yield the same results

```csharp
var match1 = list.ExceptBy(filter, x => x.State, StringComparer.OrdinalIgnoreCase).Select(x => x.Name).ToList();
var match2 = list.Where(x => !filter.Contains(x.State, StringComparer.OrdinalIgnoreCase)).Select(x => x.Name).ToList();
```

But they don't.

The query using `ExceptBy` seems to only want to return distinct states, and will the example above filter out the last item ("Should be returned 5"), returning only one item per allowed state.

I find it hard to believe this could be by design, as the statement clearly reads that the list should be filtered by the second enumerable. There's no indication anywhere that it should perform a selection on the items that are not filtered.

### Reproduction Steps

Run the code in the description

### Expected behavior

I expect `match1 `and `match2 `to yield the same results

### Actual behavior

`match1` will only yield one result per distinct state.

**Comment by KeterSCP** [comment]
Situation is the same as in https://github.com/dotnet/runtime/issues/75066
[Documentation](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.exceptby?view=net-7.0) says that method return value is

>A sequence that contains the set difference of the elements of two sequences.

Result of a set difference is also a set. And by definition set's elements are distinct.

**Comment by AbstractionsAs** [comment]
@KeterSCP That would make sense if the distnct operation was performed on the result type, but in this case it's distinct by the filtering key.

Does that make sense at all to you?

**Comment by KeterSCP** [comment]
Yes, you are performing a set operation based on specified selector key, not on the type itself

**Comment by AbstractionsAs** [comment]
@KeterSCP It doesn't at all to me. "[this] except [that] by [property]" might return unique instances of [this], but not only one instance of [this] per distinct [property]. 

If this is indeed by design, then I think this needs to be given a lot more attention in the documentation.
